Spring 02 orm

alchinYY opened the pull request:
1. Смена подхода с jdbc на orm hibernate.
2. Решена проблема n+1 для select запросов. 
3. Добавлена новая сущность - комменты к книгам
4. Переписаны тесты

Commit:
Структура БД переведана на ORM-подход

Commit:
Обновлены имена тестов

kataus commented on spring-02-library/src/main/java/ru/otus/spring/library/dao/impl/AuthorJdbcDao.java:
```diff
@@ -1,62 +1,52 @@
     @Override
     public void deleteById(Long id) {
-        jdbc.update("delete from authors where id = :id", Map.of("id", id));
+        getById(id).ifPresent(em::remove);
```
👍🏻 красиво

alchinYY replied:
спасибо!

kataus commented on spring-02-library/src/main/java/ru/otus/spring/library/dao/impl/BookJdbcDao.java:
```diff
@@ -1,121 +1,60 @@
     }
 
     @Override
     public void updateById(Long id, Book entity) {
```
Нам отдельно эта функция не очень нужна. обновление сущности можем сделать и на уровне сервиса, а для сохранения внешних данных save за глаза

alchinYY replied:
Согласен, отличная идея!

kataus commented on spring-02-library/src/main/java/ru/otus/spring/library/dao/impl/CommentsJdbcDao.java:
```diff
@@ -0,0 +1,61 @@
+
+    @Override
+    public Optional<Comment> getById(Long id) {
+        TypedQuery<Comment> query = em.createQuery("SELECT c FROM Comment c LEFT JOIN FETCH c.book where c.id = :id", Comment.class);
```
Тут  left не является удачным выбором. Можем осознано применять inner, он быстрее гораздо.Да и вообще можно всё на описании модели реализовать, нам комментарий без книги не нужен ни разу, потому можно спокойно делать EAGER связь.

kataus commented on spring-02-library/src/main/java/ru/otus/spring/library/dao/impl/CommentsJdbcDao.java:
```diff
@@ -0,0 +1,61 @@
+
+    @Override
+    public void updateById(Long id, Comment entity) {
+        Query query = em.createQuery("UPDATE Comment c SET c.body = :body WHERE c.id = :id");
```
Ни в коем случае так в ORM не делаем, не пишем явно скрипты обновления данных, только в крайнем случае, но это не он

alchinYY replied:
Полностью переписал метод

kataus commented on spring-02-library/src/main/java/ru/otus/spring/library/domain/Book.java:
```diff
@@ -1,23 +1,44 @@
     private String name;
+
+    @Fetch(FetchMode.SUBSELECT)
+    @ManyToMany(targetEntity = Author.class, fetch = FetchType.LAZY)
```
У many-to-many fetch = LAZY по умочанию, не надо повторяться, это усложняет чтение

alchinYY replied:
+

kataus commented on spring-02-library/src/main/java/ru/otus/spring/library/domain/Book.java:
```diff
@@ -1,23 +1,44 @@
+    @BatchSize(size = 1)
+    @ManyToOne(fetch = FetchType.LAZY)
+    @JoinColumn(name = "genre_id")
     private Genre genre;
```
Мы получаем хоть раз книги без авторов/жанров? Может лучше сразу написать EAGER и меньше кода?

alchinYY replied:
Запрашивается, у нас есть сущность - Комментарий. Для него нет нужны запрашивать жанры и авторов

kataus commented on spring-02-library/src/main/java/ru/otus/spring/library/domain/Comment.java:
```diff
@@ -0,0 +1,62 @@
+    @Column(name = "body", nullable = false, length = 500)
+    private String body;
+
+    @ManyToOne(fetch = FetchType.LAZY, cascade = {CascadeType.DETACH})
```
Писал выше, почему EAGER  тут более удачный выбор

alchinYY replied:
Не совсем согласен. При получении списка комментариев нам не нужна полная информация о книге. 
Достаточно только id и имя. И по этому eager можно и не трогать)

kataus replied:
Комментарии без книги мало несут пользы, это зависимая от книги сущность и без неё не имеет практического смысла. Но вопрос продукта,  логика правильная

kataus commented on spring-02-library/src/main/java/ru/otus/spring/library/domain/Comment.java:
```diff
@@ -0,0 +1,62 @@
+    @Column(name = "date_created", updatable = false)
+    private LocalDateTime date;
+
+    @Column(name = "body", nullable = false, length = 500)
```
nullable, length - имеют смысл если мы генерим код для базы из JPA. Во всех остальных случаях мы понимаем что это просто для информации

alchinYY replied:
А если в будущем планируется миграцию проводить с одного способа версионирования БД на другой + некоторая информативность для потомков кода?

kataus replied:
Не принято в проме катиться из JPA модели, только на первых стадиях прототипирования. Есть более сильные инструменты версионирования базы. Правильно что воспринимаем это как мета-информацию и не надеемся на проверки от фреймворка

kataus commented on spring-02-library/src/main/java/ru/otus/spring/library/service/impl/AuthorService.java:
```diff
@@ -15,27 +17,33 @@
     private final Dao<Long, Author> genreDao;
 
     @Override
+    @Transactional(readOnly = true)
```
Явно тут ставить транзакцию излишне, у нас всего один запрос в базу, который сам поднимет транзакцию

kataus commented on spring-02-library/src/main/java/ru/otus/spring/library/service/impl/AuthorService.java:
```diff
@@ -15,27 +17,33 @@
     @Override
+    @Transactional
     public Author updateById(Long id, Author author){
         genreDao.updateById(id, author);
```
Мы берем author из базы, обновляем у него поля и всё, по закрытию транзакции это всё уедет в базу

kataus commented on spring-02-library/src/main/java/ru/otus/spring/library/service/impl/CommentService.java:
```diff
@@ -0,0 +1,56 @@
+
+    @Transactional(readOnly = true)
+    public List<Comment> getAllByBookId(Long id) {
+        return commentDao.getByBookId(id);
```
Нам не нужен для этого отдельный метод в DAO, ORM дает достаточно возможностей реализовать это. давайте в dao этот метод удалим и без нег осделаем

alchinYY replied:
Я не смог придумать, как упростить сам запрос к БД. Иначе Lazy в моей реализации становится не очень полезным(

kataus replied:
Именно потому я этот вопрос явно и задаю. чтобы остался LAZY, при этом научиться материализовывать прокси.

Commit:
Поправлены замечания из коммита

kataus commented on spring-02-library/src/main/java/ru/otus/spring/library/dao/impl/BookJdbcDao.java:
```diff
@@ -21,20 +21,13 @@
     public Optional<Book> getById(Long id) {
         TypedQuery<Book> query = em.createQuery(
-                "SELECT b FROM Book b LEFT JOIN FETCH b.authors LEFT JOIN FETCH b.genre WHERE b.id = :id",
+                "SELECT b FROM Book b LEFT JOIN FETCH b.authors LEFT JOIN FETCH b.genre JOIN FETCH b.comments WHERE b.id = :id",
```
Давайте не так будем материализовывать прокси, а возможностями ORM. Попробуем в учебном проекте чтобы уже явно видеть в реальной работе. оно не очень очевидно

kataus commented on spring-02-library/src/main/java/ru/otus/spring/library/service/impl/AuthorService.java:
```diff
@@ -14,37 +14,36 @@
-        return getById(id);
+        Author authorFromDb = getById(id);
+        authorFromDb.setName(author.getName());
+        return authorDao.save(authorFromDb);
```
save тут тоже лишний, по закрытию транзакции изменения коммитятся в базу

Commit:
Доверился фреймворку для update сущнстей)
Во всех запросах get by id сделал без запросов, чисто на em.
Большую часть сущностей превратил в eager.

Commit:
Убрал зависимость book в comment

Commit:
Восстановил работоспособность тестов

Commit:
минификс
